### PR TITLE
HDDS-6545. OzoneManagerUtils#getBucketLayout should not suppress bucket key lookup errors.

### PR DESCRIPTION
**Pull request by JyotinderSingh**
## What changes were proposed in this pull request?

`OzoneManagerUtils#getBucketLayout` catches the bucket key lookup errors that might occur and returns BucketLayout.DEFAULT.
This behavior should be changed to throwing an error - otherwise it creates uncertainty.

This patch also reduces code duplication across the project for the `getBucketLayout` method.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6545

## How was this patch tested?

Integration Tests.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 1]
```diff
@@ -338,8 +338,8 @@ public OmKeyInfo lookupKey(OmKeyArgs args, String clientAddress)
         bucketName);
 
     BucketLayout bucketLayout =
-        getBucketLayout(metadataManager, args.getVolumeName(),
-            args.getBucketName());
+        OzoneManagerUtils.getBucketLayout(args.getVolumeName(),
+            args.getBucketName(), ozoneManager, new HashSet<>());
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java:342]: I have minor code suggestions for this refactoring:

1. Create another `getBucketLayout` method (in `OzoneManagerUtils`) without the `Set<Pair<String, String>> visited` parameter, passing `new HashSet<>()` to the existing `getBucketLayout` method.
2. Make `OzoneManager` the first param of this new method.
3. Add static import for `OzoneManagerUtils.getBucketLayout`.

This would make the code more readable and minimize the changes necessary to satisfy line width rules.

Comment (JyotinderSingh) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: Thanks a lot for the reviews @adoroszlai, I have addressed these points in my latest patch.
I had a typo in one place which seemed to be causing the CI errors, waiting on the new run to verify the behavior now.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 2]
```diff
@@ -830,7 +831,9 @@ public boolean addAcl(OzoneObj obj, OzoneAcl acl) throws IOException {
       OMFileRequest.validateBucket(metadataManager, volume, bucket);
       String objectKey = metadataManager.getOzoneKey(volume, bucket, keyName);
       OmKeyInfo keyInfo = metadataManager
-          .getKeyTable(getBucketLayout(metadataManager, volume, bucket))
+          .getKeyTable(
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManagerUtils.java [hunk 3]
```diff
@@ -121,11 +122,19 @@ public static BucketLayout getBucketLayout(String volName,
...
+    LOG.error("Could not get BucketLayout. Volume: {}, Bucket: {} ", volName,
+        buckName);
+    OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
+    OmVolumeArgs volumeArgs = omMetadataManager.getVolumeTable()
+        .get(omMetadataManager.getVolumeKey(volName));
+    if (volumeArgs == null) {
+      throw new OMException("Could not get BucketLayout. Volume not found: "
+          + volName, OMException.ResultCodes.VOLUME_NOT_FOUND);
+    }
+    throw new OMException("Could not get BucketLayout." +
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManagerUtils.java]: Thank you for the reviews @rakeshadr, I have addressed these points in my latest patch.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java [hunk 4]
```diff
@@ -2339,8 +2360,8 @@ public OmKeyInfo getPendingDeletionDir() throws IOException {
     long countEntries = 0;
 
     Table fileTable = metadataManager.getKeyTable(
-        getBucketLayout(metadataManager, parentInfo.getVolumeName(),
-            parentInfo.getBucketName()));
+        OzoneManagerUtils.getBucketLayout(parentInfo.getVolumeName(),
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java]: Thank you for the suggestion, I have added this in my latest patch.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzonePrefixPathImpl.java [hunk 5]
```diff
@@ -76,7 +76,7 @@ public OzonePrefixPathImpl(String volumeName, String bucketName,
       // set recursive access check to true if this directory contains
       // sub-directories or sub-files.
       checkRecursiveAccess = OMFileRequest.hasChildren(
-          pathStatus.getKeyInfo(), keyManager.getMetadataManager());
+          pathStatus.getKeyInfo(), keyManagerImpl.getMetadataManager());
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzonePrefixPathImpl.java:79]: Minor: Unwanted change. You can use `keyManager.getMetadataManager()`

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerRatisRequest.java [hunk 6]
```diff
@@ -58,6 +63,15 @@ public void testRequestWithNonExistentBucket()
 
     String volumeName = "vol1";
     String bucketName = "invalidBuck";
+
+    // Add entry to Bucket Table.
+    omMetadataManager.getBucketTable().addCacheEntry(
```
Comment (rakeshadr) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerRatisRequest.java:68]: Is this related to the patch?

Comment (JyotinderSingh) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerRatisRequest.java]: Yes, this test began failing due to a change in the internal implementation of getBucketLayout. So we need to add the volume to the volume table for this to work correctly.
I made a small error in this test change by adding a key to the bucket table instead, I have fixed that in my latest patch.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerRatisRequest.java [hunk 7]
```diff
@@ -58,14 +61,29 @@ public void testRequestWithNonExistentBucket()
...
-    Assert.assertNotNull(req);
-    Assert.assertTrue("Unexpected request on invalid bucket",
-        req instanceof S3MultipartUploadCompleteRequest);
+    try {
+      // Request creation flow should throw exception if the bucket associated
+      // with the request doesn't exist.
+      OzoneManagerRatisUtils.createClientRequest(omRequest, ozoneManager);
+      fail("Expected OMException: Bucket not found");
+    } catch (OMException oe) {
+      // Expected exception.
```
Comment (rakeshadr) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerRatisRequest.java:86]: Can you add an assertion inside the catch block.

      // Expected exception
      Assert.assertEquals(OMException.ResultCodes.BUCKET_NOT_FOUND,
          oe.getResult());

**Review by rakeshadr — COMMENTED** [review]
Thanks @JyotinderSingh for the updated patch. Added a minor imprv in test.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerRatisRequest.java (hunk 7 above)
Comment (JyotinderSingh) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/ratis/TestOzoneManagerRatisRequest.java:87]: Thank you for the suggestion, I've added this in my latest commit.

**Review by rakeshadr — APPROVED** [review]
+1 LGTM. Thanks @JyotinderSingh for the contribution.
